File: internal/cli/output/console.go
// Package output provides formatted console output utilities for the Xyncra CLI.
// It uses only standard library packages (text/tabwriter, encoding/json) per D-041.
package output

import (
	"encoding/json"
	"fmt"
	"io"
	"text/tabwriter"
)

// ConsoleWriter writes formatted output to an io.Writer.
type ConsoleWriter struct {
	w io.Writer
}

// KeyValueEntry represents a key-value pair for aligned display.
type KeyValueEntry struct {
	Key   string
	Value string
}

// NewConsoleWriter creates a ConsoleWriter that writes to w.
func NewConsoleWriter(w io.Writer) *ConsoleWriter {
	return &ConsoleWriter{w: w}
}

// Table writes a tabular output with headers and rows using text/tabwriter.
// Headers are printed first, followed by a separator line (---), then the data rows.
// If rows is empty, only headers and the separator line are printed.
func (cw *ConsoleWriter) Table(headers []string, rows [][]string) {
	tw := tabwriter.NewWriter(cw.w, 0, 0, 2, ' ', 0)

	// Print headers separated by tabs.
	for i, h := range headers {
		if i > 0 {
			fmt.Fprint(tw, "\t")
		}
		fmt.Fprint(tw, h)
	}
	fmt.Fprintln(tw)

	// Print separator line.
	for i, h := range headers {
		if i > 0 {
			fmt.Fprint(tw, "\t")
		}
		for j := 0; j < len(h); j++ {
			fmt.Fprint(tw, "-")
		}
	}
	fmt.Fprintln(tw)

	// Print data rows.
	for _, row := range rows {
		for i, cell := range row {
			if i > 0 {
				fmt.Fprint(tw, "\t")
			}
			fmt.Fprint(tw, cell)
		}
		fmt.Fprintln(tw)
	}

	tw.Flush()
}

// KeyValue writes aligned key-value pairs. Keys are left-aligned and padded
// to the length of the longest key, followed by ": " and the value.
func (cw *ConsoleWriter) KeyValue(pairs []KeyValueEntry) {
	if len(pairs) == 0 {
		return
	}

	// Find the longest key length for alignment.
	maxKeyLen := 0
	for _, p := range pairs {
		if len(p.Key) > maxKeyLen {
			maxKeyLen = len(p.Key)
		}
	}

	for _, p := range pairs {
		fmt.Fprintf(cw.w, "%-*s  %s\n", maxKeyLen, p.Key, p.Value)
	}
}

// JSONPretty writes v as indented JSON to the underlying writer, followed by a newline.
// It returns an error if JSON marshalling fails.
func (cw *ConsoleWriter) JSONPretty(v any) error {
	data, err := json.MarshalIndent(v, "", "  ")
	if err != nil {
		return err
	}
	_, err = cw.w.Write(data)
	if err != nil {
		return err
	}
	_, err = fmt.Fprintln(cw.w)
	return err
}

// Section writes a section title followed by a newline.
func (cw *ConsoleWriter) Section(title string) {
	fmt.Fprintln(cw.w, title)
}

// Info writes an informational message followed by a newline.
func (cw *ConsoleWriter) Info(msg string) {
	fmt.Fprintln(cw.w, msg)
}

// Error writes an error message in the format "Error: msg\n".
func (cw *ConsoleWriter) Error(msg string) {
	fmt.Fprintf(cw.w, "Error: %s\n", msg)
}

File: demo/web/packages/xyncra-client-cli/src/output/console.ts
/**
 * Console output writer with tabwriter-equivalent alignment.
 *
 * Mirrors Go internal/cli/output/console.go exactly.
 *
 * @module
 */

import type { Writable } from 'node:stream';

/** Key-value pair for aligned display. */
export interface KeyValueEntry {
  key: string;
  value: string;
}

/**
 * ConsoleWriter writes formatted output to a Writable stream.
 * Mirrors Go ConsoleWriter.
 */
export class ConsoleWriter {
  private readonly w: Writable;

  constructor(w: Writable = process.stdout) {
    this.w = w;
  }

  /**
   * Write a table with headers and rows. Uses manual column-width alignment
   * (equivalent to Go text/tabwriter with minwidth=0, tabwidth=0, padding=2).
   */
  table(headers: string[], rows: string[][]): void {
    // Calculate column widths.
    const colWidths = headers.map((h) => h.length);
    for (const row of rows) {
      for (let i = 0; i < row.length; i++) {
        if (i < colWidths.length) {
          colWidths[i] = Math.max(colWidths[i], row[i].length);
        }
      }
    }

    const pad = (s: string, w: number): string => s + ' '.repeat(Math.max(0, w - s.length));

    // Print headers.
    const headerLine = headers.map((h, i) => pad(h, colWidths[i])).join('  ');
    this.w.write(headerLine + '\n');

    // Print separator.
    const sepLine = headers.map((h, i) => '-'.repeat(h.length) + (i < headers.length - 1 ? '  ' : '')).join('');
    // Simpler: just use dashes with same column widths
    const sepLine2 = colWidths.map((w, i) => '-'.repeat(w) + (i < colWidths.length - 1 ? '  ' : '')).join('');
    this.w.write(sepLine2 + '\n');

    // Print rows.
    for (const row of rows) {
      const line = row.map((cell, i) => {
        const w = i < colWidths.length ? colWidths[i] : cell.length;
        return pad(cell, w);
      }).join('  ');
      this.w.write(line + '\n');
    }
  }

  /** Write aligned key-value pairs. */
  keyValue(pairs: KeyValueEntry[]): void {
    if (pairs.length === 0) return;
    const maxKeyLen = Math.max(...pairs.map((p) => p.key.length));
    for (const p of pairs) {
      this.w.write(`${p.key.padEnd(maxKeyLen)}  ${p.value}\n`);
    }
  }

  /** Write v as indented JSON. */
  jsonPretty(v: unknown): void {
    this.w.write(JSON.stringify(v, null, 2) + '\n');
  }

  /** Write a section title followed by a newline. */
  section(title: string): void {
    this.w.write(title + '\n');
  }

  /** Write an informational message. */
  info(msg: string): void {
    this.w.write(msg + '\n');
  }

  /** Write an error message. */
  error(msg: string): void {
    this.w.write(`Error: ${msg}\n`);
  }
}
